docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,649 @@
+  CRDT implementations in future.
+
+
+Each object is also created with a JSON schema. The schema is represented by an
+initial `schema.json` and a series of schema migrations which extend that
+initial schema. Schema migrations will not be addressed in detail in this RFC
+but we will show their feasibility.
+
+#### `schema`
+
+Schemas are primarily important for the interoperability of the system. We need
+applications to be able to rely on the data they are working with being valid,
+otherwise we impose the problem of schema validation on application developers.
+We represent schemas using JSON schema which is in the `schema/schema.json` of any
+object tree.
+
+Schemas will need to be able to change, the `schema/migrations` directory is
+present to allow us to store compatible changes to a schema in future. Schema
+migration is out of scope for this RFC.
+
+
+#### `change`
+
+This is the automerge change which this commit introduces. It is a binary file
+which must contain a single change and it's dependents must be the dependents
+referenced by the parents of the commit.
+
+
+#### `author`
+
+This is a file containing the SHA of the commit which references the authors
+identity. We need this so we know to ignore this commit when walking the
+history to look for changes.
+
+#### `signatures`
+
+The commit is signed using the same mechanism as identity documents. Namely
+the `X-rad-signature` trailer.
+
+### Reconstructing Collaborative Objects
+
+Assuming we have replicated a number of collaborative objects from our tracking
+graph, we can now view the merged state of those objects. To do this we search
+through every `/rad/collaborative-objects/<typename>/<object ID>` reference for
+every remote we have and collect the change files for each object ID.
+
+At this point we have the hash linked graph of automerge changes, but we need
+to make sure that the merged document is authenticated and valid with respect
+to it's schema. To do this we start at the roots of the hash graph and walk
+down the tree. As we encounter each change we check it's signature, apply it
+and check that the new document does not violate the schema. If it does violate
+the schema we discard the change and all dependent changes. Finally, we have an
+authenticated document which respects a given schema.
+
+It is important to note that this merging is at this point not stored in the
+repository - it can be performed in memory and may be cached. The result is
+that the user sees a single merged view of the object based on the contents
+of the remotes they have replicated. That is, there is no additional
+merge-then-commit step.
+
+### Fetching Collaborative Objects
+
+Each time a repository creates a collaborative object tree it creates a ref
+pointing to that object at `refs/namespaces/<project>/rad/collaborative-objects/<typename>/<object ID>`, 
+where `object ID` is a unique identifier generated at creation time. 
```

Comment:
Is this the UUID in the `manifest`? I'm not sure if that is explicitly stated anywhere

Reply:
Added a reference.